# Clean & Transform Census Datasets

Importing, cleaning and combining census datasets of 2011, 2016 and 2021. Census data for each year contains population by age, sex and personal income.

Datasets for each census year contains 3 separate datasets; population by age, gender and income. These datasets are first individually cleaned and combined to create a 1 dataset for each year.

In [30]:
import pandas as pd
import numpy as np
import os

## Import & Combine 2011 Census Datasets

### Import & Clean 2011 Population-Age Dataset

In [31]:
# set file path and import 2011 population by age dataset
path = r"/Users/patel/Documents/CF-Data Anaylst Course/portfolio_projects/mbs_analysis/datasets/"

df_census_age_2011 = pd.read_csv(
    os.path.join(path, "original_datasets/census_data/2011/2011_population_age.csv"),
    index_col=False,
    encoding="ISO-8859-1",
)
df_census_age_2011.head()

SA3  age_0-14  age_15-24  age_25-44  age_45-64  age_65-79  age_80+
0  10101     13648       7590      15585      18904       8708     3072
1  10102     11025       6762      15554      14760       4171     1239
2  10103      3641       2198       4485       5530       2264      818
3  10104     11507       6347      12033      21878      11980     4123
4  10201     30502      19699      37671      43996      20940    10297

In [32]:
df_census_age_2011.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 351 entries, 0 to 350
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   SA3        351 non-null    int64
 1   age_0-14   351 non-null    int64
 2   age_15-24  351 non-null    int64
 3   age_25-44  351 non-null    int64
 4   age_45-64  351 non-null    int64
 5   age_65-79  351 non-null    int64
 6   age_80+    351 non-null    int64
dtypes: int64(7)
memory usage: 19.3 KB


In [33]:
# converting SA3 to be string for merges with other datasets (census and mbs)
df_census_age_2011["SA3"] = df_census_age_2011["SA3"].astype(("str"))
df_census_age_2011.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 351 entries, 0 to 350
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   SA3        351 non-null    object
 1   age_0-14   351 non-null    int64 
 2   age_15-24  351 non-null    int64 
 3   age_25-44  351 non-null    int64 
 4   age_45-64  351 non-null    int64 
 5   age_65-79  351 non-null    int64 
 6   age_80+    351 non-null    int64 
dtypes: int64(6), object(1)
memory usage: 19.3+ KB


In [34]:
df_census_age_2011.describe()

age_0-14     age_15-24     age_25-44     age_45-64     age_65-79  \
count    351.000000    351.000000    351.000000    351.000000    351.000000   
mean   11806.190883   8166.660969  17179.524217  15541.148148   6189.886040   
std     8339.142607   6068.821251  13064.111914  10225.126635   4261.153964   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%     6461.000000   4083.500000   8002.000000   8539.500000   3072.500000   
50%     9796.000000   6702.000000  13932.000000  13716.000000   5595.000000   
75%    15842.500000  10846.000000  23419.000000  21342.500000   8696.500000   
max    39077.000000  28164.000000  87146.000000  45569.000000  22249.000000   

            age_80+  
count    351.000000  
mean    2392.233618  
std     1902.499656  
min        0.000000  
25%     1019.500000  
50%     1987.000000  
75%     3215.000000  
max    10297.000000

Total 351 SA3 areas in census population age dataset. This is correct since MBS 2013-19 contains 346 unique SA3 areas. Further checks on missing SA3 areas will be conducted before merging datasets

### Import & Clean 2011 Population Personal Income Dataset

In [35]:
# set file path and import 2011 population by age dataset
path = r"/Users/patel/Documents/CF-Data Anaylst Course/portfolio_projects/mbs_analysis/datasets/"

df_census_personal_income_2011 = pd.read_csv(
    os.path.join(
        path, "original_datasets/census_data/2011/2011_population_personal_income.csv"
    ),
    index_col=False,
    encoding="ISO-8859-1",
)
df_census_personal_income_2011.head()

SA3  Negative income  Nil income  $1-$199 ($1-$10,399)  \
0  10101              333        3126                  3900   
1  10102              174        2503                  2489   
2  10103               83         799                  1103   
3  10104              295        2846                  4480   
4  10201              640        8641                 10113   

   $200-$299 ($10,400-$15,599)  $300-$399 ($15,600-$20,799)  \
0                         6157                         6630   
1                         2660                         3051   
2                         1479                         1746   
3                         8533                         8105   
4                        13944                        15751   

   $400-$599 ($20,800-$31,199)  $600-$799 ($31,200-$41,599)  \
0                         7080                         5958   
1                         3978                         4239   
2                         2064                         1878   
3                         9082                         6388   
4                        17655                        13977   

   $800-$999 ($41,600-$51,999)  $1,000-$1,249 ($52,000-$64,999)  \
0                         4256                             3934   
1                         3859                             4422   
2                         1344                             1134   
3                         4243                             3191   
4                        10481                             9723   

   $1,250-$1,499 ($65,000-$77,999)  $1,500-$1,999 ($78,000-$103,999)  \
0                             2656                              3222   
1                             3528                              4495   
2                              695                               787   
3                             1679                              1965   
4                             6587                              8093   

   $2,000 or more ($104,000 or more)  Not stated  Not applicable   Total  
0                               2258        4338           13648   67500  
1                               4085        3007           11025   53511  
2                                530        1642            3641   18931  
3                               1123        4432           11507   67880  
4                               6307       10701           30502  163110

In [36]:
df_census_personal_income_2011.shape

(351, 16)

In [37]:
df_census_personal_income_2011.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 351 entries, 0 to 350
Data columns (total 16 columns):
 #   Column                             Non-Null Count  Dtype
---  ------                             --------------  -----
 0   SA3                                351 non-null    int64
 1   Negative income                    351 non-null    int64
 2   Nil income                         351 non-null    int64
 3   $1-$199 ($1-$10,399)               351 non-null    int64
 4   $200-$299 ($10,400-$15,599)        351 non-null    int64
 5   $300-$399 ($15,600-$20,799)        351 non-null    int64
 6   $400-$599 ($20,800-$31,199)        351 non-null    int64
 7   $600-$799 ($31,200-$41,599)        351 non-null    int64
 8   $800-$999 ($41,600-$51,999)        351 non-null    int64
 9   $1,000-$1,249 ($52,000-$64,999)    351 non-null    int64
 10  $1,250-$1,499 ($65,000-$77,999)    351 non-null    int64
 11  $1,500-$1,999 ($78,000-$103,999)   351 non-null    int64
 12  $2,000 or more ($104,0

In [38]:
# converting SA3 to be string for merges with other datasets (census and mbs)
df_census_personal_income_2011["SA3"] = df_census_age_2011["SA3"].astype(("str"))
df_census_personal_income_2011.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 351 entries, 0 to 350
Data columns (total 16 columns):
 #   Column                             Non-Null Count  Dtype 
---  ------                             --------------  ----- 
 0   SA3                                351 non-null    object
 1   Negative income                    351 non-null    int64 
 2   Nil income                         351 non-null    int64 
 3   $1-$199 ($1-$10,399)               351 non-null    int64 
 4   $200-$299 ($10,400-$15,599)        351 non-null    int64 
 5   $300-$399 ($15,600-$20,799)        351 non-null    int64 
 6   $400-$599 ($20,800-$31,199)        351 non-null    int64 
 7   $600-$799 ($31,200-$41,599)        351 non-null    int64 
 8   $800-$999 ($41,600-$51,999)        351 non-null    int64 
 9   $1,000-$1,249 ($52,000-$64,999)    351 non-null    int64 
 10  $1,250-$1,499 ($65,000-$77,999)    351 non-null    int64 
 11  $1,500-$1,999 ($78,000-$103,999)   351 non-null    int64 
 12  $2,000 o

In [39]:
df_census_personal_income_2011.describe()

Negative income    Nil income  $1-$199 ($1-$10,399)  \
count       351.000000    351.000000            351.000000   
mean        290.849003   3744.478632           3662.404558   
std         215.583083   3165.936958           2641.206295   
min           0.000000      0.000000              0.000000   
25%         146.000000   1678.000000           1863.000000   
50%         245.000000   2844.000000           3079.000000   
75%         377.000000   4752.500000           4871.000000   
max        1211.000000  16505.000000          13276.000000   

       $200-$299 ($10,400-$15,599)  $300-$399 ($15,600-$20,799)  \
count                   351.000000                   351.000000   
mean                   5120.683761                  4888.470085   
std                    3776.026852                  3382.369116   
min                       0.000000                     0.000000   
25%                    2451.000000                  2439.500000   
50%                    4415.000000                  4464.000000   
75%                    7091.500000                  6687.000000   
max                   23608.000000                 16338.000000   

       $400-$599 ($20,800-$31,199)  $600-$799 ($31,200-$41,599)  \
count                   351.000000                   351.000000   
mean                   5715.242165                  5127.193732   
std                    3862.770369                  3568.099756   
min                       0.000000                     0.000000   
25%                    2989.000000                  2527.000000   
50%                    5119.000000                  4493.000000   
75%                    7828.000000                  6700.500000   
max                   18110.000000                 16934.000000   

       $800-$999 ($41,600-$51,999)  $1,000-$1,249 ($52,000-$64,999)  \
count                   351.000000                       351.000000   
mean                   4092.131054                      3908.675214   
std                    2917.502148                      2842.593911   
min                       0.000000                         0.000000   
25%                    2109.500000                      1919.000000   
50%                    3610.000000                      3329.000000   
75%                    5469.000000                      5286.500000   
max                   12938.000000                     13685.000000   

       $1,250-$1,499 ($65,000-$77,999)  $1,500-$1,999 ($78,000-$103,999)  \
count                       351.000000                        351.000000   
mean                       2733.763533                       3193.450142   
std                        2061.148821                       2553.227144   
min                           0.000000                          0.000000   
25%                        1292.500000                       1402.000000   
50%                        2247.000000                       2552.000000   
75%                        3795.500000                       4480.000000   
max                       11634.000000                      15848.000000   

       $2,000 or more ($104,000 or more)    Not stated  Not applicable  \
count                         351.000000    351.000000      351.000000   
mean                         3080.592593   3910.615385    11806.193732   
std                          3521.879540   2887.139860     8339.139214   
min                             0.000000      0.000000        0.000000   
25%                           871.500000   1997.500000     6461.000000   
50%                          1936.000000   3264.000000     9796.000000   
75%                          4029.500000   5182.000000    15842.500000   
max                         21922.000000  22876.000000    39077.000000   

               Total  
count     351.000000  
mean    61275.410256  
std     41775.190657  
min         0.000000  
25%     33165.500000  
50%     53373.000000  
75%     82641.000000  
max    178432.000000

In [40]:
df_census_personal_income_2011.drop(["Total"], axis=1, inplace=True)
df_census_personal_income_2011.columns

Index(['SA3', 'Negative income', 'Nil income', '$1-$199 ($1-$10,399)',
       '$200-$299 ($10,400-$15,599)', '$300-$399 ($15,600-$20,799)',
       '$400-$599 ($20,800-$31,199)', '$600-$799 ($31,200-$41,599)',
       '$800-$999 ($41,600-$51,999)', '$1,000-$1,249 ($52,000-$64,999)',
       '$1,250-$1,499 ($65,000-$77,999)', '$1,500-$1,999 ($78,000-$103,999)',
       '$2,000 or more ($104,000 or more)', 'Not stated', 'Not applicable'],
      dtype='object')

In [41]:
rename_columns = {
    "$1-$199 ($1-$10,399)": "$1-$10,399",
    "$200-$299 ($10,400-$15,599)": "$10,400-$15,599",
    "$300-$399 ($15,600-$20,799)": "$15,600-$20,799",
    "$400-$599 ($20,800-$31,199)": "$20,800-$31,199",
    "$600-$799 ($31,200-$41,599)": "$31,200-$41,599",
    "$800-$999 ($41,600-$51,999)": "$41,600-$51,999",
    "$1,000-$1,249 ($52,000-$64,999)": "$52,000-$64,999",
    "$1,250-$1,499 ($65,000-$77,999)": "$65,000-$77,999",
    "$1,500-$1,999 ($78,000-$103,999)": "$78,000-$103,999",
    "$2,000 or more ($104,000 or more)": "$104,000+",
}

df_census_personal_income_2011.rename(columns=rename_columns, inplace=True)
df_census_personal_income_2011

SA3  Negative income  Nil income  $1-$10,399  $10,400-$15,599  \
0    10101              333        3126        3900             6157   
1    10102              174        2503        2489             2660   
2    10103               83         799        1103             1479   
3    10104              295        2846        4480             8533   
4    10201              640        8641       10113            13944   
..     ...              ...         ...         ...              ...   
346  90101                3          68          27               39   
347  90102                8          35          27               49   
348  90103                0          20          15               44   
349  99797                0           0           0                0   
350  99999                0           0           0                0   

     $15,600-$20,799  $20,800-$31,199  $31,200-$41,599  $41,600-$51,999  \
0               6630             7080             5958             4256   
1               3051             3978             4239             3859   
2               1746             2064             1878             1344   
3               8105             9082             6388             4243   
4              15751            17655            13977            10481   
..               ...              ...              ...              ...   
346               38               50              109              128   
347               45               71               47               21   
348               30               27               22               22   
349                0                0                0                0   
350                0                0                0                4   

     $52,000-$64,999  $65,000-$77,999  $78,000-$103,999  $104,000+  \
0               3934             2656              3222       2258   
1               4422             3528              4495       4085   
2               1134              695               787        530   
3               3191             1679              1965       1123   
4               9723             6587              8093       6307   
..               ...              ...               ...        ...   
346              181              110               116         88   
347               24               17                23         18   
348               15               22                20         10   
349                0                0                 0          0   
350                3                3                13          0   

     Not stated  Not applicable  
0          4338           13648  
1          3007           11025  
2          1642            3641  
3          4432           11507  
4         10701           30502  
..          ...             ...  
346         834             263  
347          36             129  
348          26             103  
349           0               0  
350           3               0  

[351 rows x 15 columns]

### Import & Clean 2011 Population-Gender Dataset

In [42]:
# import 2011 population by gender dataset

df_census_gender_2011 = pd.read_csv(
    os.path.join(path, "original_datasets/census_data/2011/2011_population_sex.csv"),
    index_col=False,
    encoding="ISO-8859-1",
)
df_census_gender_2011.head()

SA3  male_pop  female_pop  total_population
0  10101     33658       33841             67500
1  10102     26852       26660             53511
2  10103      9736        9193             18931
3  10104     33368       34514             67880
4  10201     78677       84435            163110

In [43]:
df_census_gender_2011.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 351 entries, 0 to 350
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   SA3               351 non-null    int64
 1   male_pop          351 non-null    int64
 2   female_pop        351 non-null    int64
 3   total_population  351 non-null    int64
dtypes: int64(4)
memory usage: 11.1 KB


In [45]:
df_census_gender_2011["SA3"] = df_census_gender_2011["SA3"].astype("str")
df_census_gender_2011["SA3"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 351 entries, 0 to 350
Series name: SA3
Non-Null Count  Dtype 
--------------  ----- 
351 non-null    object
dtypes: object(1)
memory usage: 2.9+ KB


In [46]:
df_census_gender_2011.describe()

male_pop    female_pop  total_population
count    351.000000    351.000000        351.000000
mean   30296.333333  30979.341880      61275.410256
std    20556.012001  21249.844061      41775.190657
min        0.000000      0.000000          0.000000
25%    16518.500000  16347.500000      33165.500000
50%    26221.000000  26755.000000      53373.000000
75%    40762.000000  42135.000000      82641.000000
max    92088.000000  90795.000000     178432.000000

### Combine 2011 Age, Gender, Income Datasets

In [49]:
# check that all datasets have the same SA3 values

# extracted SA3 values in to a set (set will remove duplicates if any)
sa3_age = set(df_census_age_2011["SA3"])
sa3_income = set(df_census_personal_income_2011["SA3"])
sa3_gender = set(df_census_gender_2011["SA3"])

# checks if sets are equal. expected diff_sa3 to be False, if they don't contain the same values
diff_sa3 = sa3_age == sa3_income == sa3_gender
print(diff_sa3)

True


In [50]:
df_census_age_income_2011 = df_census_age_2011.merge(
    df_census_personal_income_2011, how="inner", on="SA3"
)
df_census_combined_2011 = df_census_age_income_2011.merge(
    df_census_gender_2011, how="inner", on="SA3"
)
df_census_combined_2011.shape

(351, 24)

In [51]:
df_census_combined_2011.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 351 entries, 0 to 350
Data columns (total 24 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   SA3               351 non-null    object
 1   age_0-14          351 non-null    int64 
 2   age_15-24         351 non-null    int64 
 3   age_25-44         351 non-null    int64 
 4   age_45-64         351 non-null    int64 
 5   age_65-79         351 non-null    int64 
 6   age_80+           351 non-null    int64 
 7   Negative income   351 non-null    int64 
 8   Nil income        351 non-null    int64 
 9   $1-$10,399        351 non-null    int64 
 10  $10,400-$15,599   351 non-null    int64 
 11  $15,600-$20,799   351 non-null    int64 
 12  $20,800-$31,199   351 non-null    int64 
 13  $31,200-$41,599   351 non-null    int64 
 14  $41,600-$51,999   351 non-null    int64 
 15  $52,000-$64,999   351 non-null    int64 
 16  $65,000-$77,999   351 non-null    int64 
 17  $78,000-$103,999

In [53]:
df_census_combined_2011.describe().to_clipboard()